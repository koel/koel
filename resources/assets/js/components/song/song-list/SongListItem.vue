<template>
  <div>
    <h4
      v-if="showDisc && item.playable.disc"
      class="title text-k-text-primary !flex gap-2 p-2 uppercase pl-5"
    >
      Disc {{ item.playable.disc }}
    </h4>

    <article
      :class="{ playing, external, selected: item.selected }"
      class="song-item group pl-5 text-k-text-secondary border-b border-k-border !max-w-full h-[64px] flex
        items-center transition-[background-color,_box-shadow] ease-in-out duration-200
        focus:rounded-md focus focus-within:rounded-md focus:ring-inset focus:ring-1 focus:!ring-k-accent
        focus-within:ring-inset focus-within:ring-1 focus-within:!ring-k-accent
        hover:bg-white/5 hover:ring-inset hover:ring-1 hover:ring-white/10 hover:rounded-md"
      data-testid="song-item"
      tabindex="0"
      @dblclick.prevent.stop="play"
    >
      <span v-if="shouldShowColumn('track')" class="track-number">
        <SoundBars v-if="playable.playback_state === 'Playing'" />
        <span v-else class="text-k-text-secondary">
          <template v-if="isSong(playable)">{{ playable.track || '' }}</template>
          <Icon v-else :icon="faPodcast" />
        </span>
      </span>
      <span class="thumbnail leading-none">
        <SongThumbnail :playable="playable" />
      </span>
      <span class="title-artist flex flex-col gap-2 overflow-hidden">
        <span class="title text-k-text-primary !flex gap-2 items-center">
          <ExternalMark v-if="external" class="!inline-block" />
          {{ playable.title }}
        </span>
        <span class="artist">{{ artist }}</span>
      </span>
      <span v-if="shouldShowColumn('album')" class="album">{{ album }}</span>
      <template v-if="config.collaborative">
        <span class="collaborator">
          <UserAvatar :user="collaborator" width="24" />
        </span>
        <span :title="playable.collaboration.added_at" class="added-at">{{ playable.collaboration.fmt_added_at }}</span>
      </template>
      <span v-if="shouldShowColumn('duration')" class="time">{{ fmtLength }}</span>
      <span class="extra">
        <LikeButton :playable="playable" />
      </span>
    </article>
  </div>
</template>

<script lang="ts" setup>
import { faPodcast } from '@fortawesome/free-solid-svg-icons'
import { computed, toRefs, withDefaults } from 'vue'
import { getPlayableProp, requireInjection } from '@/utils/helpers'
import { isSong } from '@/utils/typeGuards'
import { secondsToHis } from '@/utils/formatters'
import { useAuthorization } from '@/composables/useAuthorization'
import { useKoelPlus } from '@/composables/useKoelPlus'
import { usePlayableListColumnVisibility } from '@/composables/usePlayableListColumnVisibility'
import { PlayableListConfigKey } from '@/symbols'

import LikeButton from '@/components/song/SongLikeButton.vue'
import SoundBars from '@/components/ui/SoundBars.vue'
import SongThumbnail from '@/components/song/SongThumbnail.vue'
import UserAvatar from '@/components/user/UserAvatar.vue'
import ExternalMark from '@/components/ui/ExternalMark.vue'

const props = withDefaults(defineProps<{ item: PlayableRow, showDisc: boolean }>(), {
  showDisc: false,
})

const emit = defineEmits<{ (e: 'play', playable: Playable): void }>()

const [config] = requireInjection<[Partial<PlayableListConfig>]>(PlayableListConfigKey, [{}])

const { currentUser } = useAuthorization()
const { isPlus } = useKoelPlus()
const { shouldShowColumn } = usePlayableListColumnVisibility()

const { item } = toRefs(props)

const playable = computed<Playable | CollaborativeSong>(() => item.value.playable)
const playing = computed(() => ['Playing', 'Paused'].includes(playable.value.playback_state!))

const external = computed(() => {
  if (!isSong(playable.value)) {
    return false
  }
  return isPlus.value && playable.value.owner_id !== currentUser.value?.id
})

const fmtLength = secondsToHis(playable.value.length)
const artist = computed(() => getPlayableProp(playable.value, 'artist_name', 'podcast_author'))
const album = computed(() => getPlayableProp(playable.value, 'album_name', 'podcast_title'))

const collaborator = computed<Pick<User, 'name' | 'avatar'>>(
  () => (playable.value as CollaborativeSong).collaboration.user,
)

const play = () => emit('play', playable.value)
</script>

<style lang="postcss" scoped>
article {
  &.droppable {
    @apply relative transition-none after:absolute after:w-full after:h-[3px] after:rounded after:bg-k-success after:top-0;

    &.dragover-bottom {
      @apply after:top-auto after:bottom-0;
    }
  }

  &.selected {
    @apply bg-white/10;
  }

  &.playing {
    .title,
    .track-number,
    .favorite {
      @apply text-k-accent !important;
    }
  }

  .title-artist {
    span {
      @apply overflow-hidden whitespace-nowrap text-ellipsis block;
    }
  }

  button {
    @apply text-current;
  }
}
</style>
