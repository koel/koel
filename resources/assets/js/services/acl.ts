import { http } from '@/services/http'
import { cache } from '@/services/cache'

type ResourceType = 'album' | 'artist' | 'radio-station' | 'user'
type Action = 'edit' | 'delete'

export const acl = {
  checkResourcePermission: async (type: ResourceType, id: string | number, action: Action) => {
    return await cache.remember(['permission', type, id, action], async () => {
      const { allowed } = await http.silently.get<{ allowed: boolean }>(`acl/permissions/${type}/${id}/${action}`)

      return allowed
    })
  },

  fetchAssignableRoles: async () => {
    return await cache.remember(['assignable-roles'], async () => {
      const { roles } = await http.get<{
        roles: Array<{ id: Role, label: string, description: string }>
      }>('acl/assignable-roles')

      return roles
    })
  },
}
