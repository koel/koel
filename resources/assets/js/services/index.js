export * from "./info";
export * from "./download";
export * from "./http";
export * from "./ls";
export * from "./playback";
export * from "./youtube";
export * from "./socket";
