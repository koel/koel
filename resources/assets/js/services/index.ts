export * from './http'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
export * from './mediaInfoService'
export * from './cache'
export * from './socketListener'
export * from './volumeManager'
