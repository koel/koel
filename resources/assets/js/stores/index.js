export * from "./album";
export * from "./artist";
export * from "./equalizer";
export * from "./favorite";
export * from "./playlist";
export * from "./preference";
export * from "./queue";
export * from "./setting";
export * from "./shared";
export * from "./song";
export * from "./user";
