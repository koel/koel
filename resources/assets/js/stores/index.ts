export * from './albumStore'
export * from './artistStore'
export * from './commonStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './overviewStore'
export * from './playlistStore'
export * from './playlistFolderStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './searchStore'
export * from './settingStore'
export * from './songStore'
export * from './themeStore'
export * from './userStore'
export * from './genreStore'
export * from './visualizerStore'
