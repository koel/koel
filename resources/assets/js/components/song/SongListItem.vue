<template>
  <div
    :class="{ playing, selected: item.selected }"
    class="song-item"
    data-testid="song-item"
    @dblclick.prevent.stop="play"
    tabindex="0"
  >
    <span v-if="columns.includes('track')" class="track-number">
      <SoundBars v-if="song.playback_state === 'Playing'"/>
      <span class="text-secondary" v-else>{{ song.track || '' }}</span>
    </span>
    <span class="thumbnail">
      <SongThumbnail :song="song"/>
    </span>
    <span v-if="columns.includes('title')" class="title text-primary">{{ song.title }}</span>
    <span v-if="columns.includes('artist')" class="artist">{{ song.artist_name }}</span>
    <span v-if="columns.includes('album')" class="album">{{ song.album_name }}</span>
    <span v-if="columns.includes('length')" class="time">{{ fmtLength }}</span>
    <span class="favorite">
      <LikeButton :song="song"/>
    </span>
  </div>
</template>

<script lang="ts" setup>
import { computed, toRefs } from 'vue'
import { playbackService } from '@/services'
import { queueStore } from '@/stores'
import { secondsToHis } from '@/utils'

import LikeButton from '@/components/song/SongLikeButton.vue'
import SoundBars from '@/components/ui/SoundBars.vue'
import SongThumbnail from '@/components/song/SongThumbnail.vue'

const props = defineProps<{ item: SongRow, columns: SongListColumn[] }>()
const { item, columns } = toRefs(props)

const song = computed(() => item.value.song)
const playing = computed(() => ['Playing', 'Paused'].includes(song.value.playback_state!))
const fmtLength = secondsToHis(song.value.length)

const play = () => {
  queueStore.queueIfNotQueued(song.value)
  playbackService.play(song.value)
}
</script>

<style lang="scss">
.song-item {
  color: var(--color-text-secondary);
  border-bottom: 1px solid var(--color-bg-secondary);
  max-width: 100% !important; // overriding .item
  height: 64px;
  display: flex;
  align-items: center;

  &:focus, &:focus-within {
    box-shadow: 0 0 1px 1px var(--color-accent) inset;
    border-radius: 4px;
  }

  @media (hover: none) {
    .cover {
      .control {
        display: flex;
      }

      &::before {
        opacity: .7;
      }
    }
  }

  &:hover {
    background: rgba(255, 255, 255, .05);
  }

  &.selected {
    background-color: rgba(255, 255, 255, .08);
  }

  &.playing {
    color: var(--color-accent);

    .title {
      color: var(--color-accent) !important;
    }
  }

  button {
    color: currentColor;
  }
}
</style>
