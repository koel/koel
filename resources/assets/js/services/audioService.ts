import { equalizerStore } from '@/stores'
import { frequencies } from '@/config'
import { dbToGain } from '@/utils'

interface Band {
  label: string
  filter: BiquadFilterNode
  db: number
}

export const audioService = {
  unlocked: false,

  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,
  preampGainNode: null as unknown as GainNode,
  analyzer: null as unknown as AnalyserNode,

  bands: [] as Band[],

  init (mediaElement: HTMLMediaElement) {
    this.element = mediaElement

    this.context = new AudioContext()
    this.preampGainNode = this.context.createGain()
    this.source = this.context.createMediaElementSource(this.element)
    this.analyzer = this.context.createAnalyser()

    this.source.connect(this.preampGainNode)

    const config = equalizerStore.getConfig()

    this.changePreampGain(config.preamp)

    let prevFilter: BiquadFilterNode

    // Create 10 bands with the frequencies similar to those of Winamp and connect them together.
    frequencies.forEach((frequency, i) => {
      const filter = this.context.createBiquadFilter()

      if (i === 0) {
        filter.type = 'lowshelf'
      } else if (i === frequencies.length - 1) {
        filter.type = 'highshelf'
      } else {
        filter.type = 'peaking'
      }

      filter.Q.setTargetAtTime(1, this.context.currentTime, 0.01)
      filter.frequency.setTargetAtTime(frequency, this.context.currentTime, 0.01)
      filter.gain.value = dbToGain(config.gains[i])

      prevFilter ? prevFilter.connect(filter) : this.preampGainNode.connect(filter)
      prevFilter = filter

      this.bands.push({
        filter,
        label: String(frequency).replace('000', 'K'),
        db: config.gains[i]
      })
    })

    prevFilter!.connect(this.analyzer)

    // connect the analyzer node last, so that changes to the equalizer affect the visualizer as well
    this.analyzer.connect(this.context.destination)

    this.unlockAudioContext()
  },

  changePreampGain (db: number) {
    this.preampGainNode.gain.value = dbToGain(db)
  },

  changeFilterGain (node: BiquadFilterNode, db: number) {
    this.bands.find(band => band.filter === node)!.db = db
    node.gain.value = dbToGain(db)
  },

  /**
   * Attempt to unlock the audio context on mobile devices by creating and playing a silent buffer upon the
   * first user interaction.
   */
  unlockAudioContext () {
    ['touchend', 'touchstart', 'click'].forEach(event => {
      document.addEventListener(event, () => {
        if (this.unlocked) return

        const source = this.context.createBufferSource()
        source.buffer = this.context.createBuffer(1, 1, 22050)
        source.connect(this.context.destination)
        source.start(0)

        this.unlocked = true
      }, {
        once: true
      })
    })
  }
}
