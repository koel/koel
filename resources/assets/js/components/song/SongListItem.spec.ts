import { expect, it } from 'vitest'
import factory from '@/__tests__/factory'
import { queueStore } from '@/stores'
import { playbackService } from '@/services'
import { screen } from '@testing-library/vue'
import UnitTestCase from '@/__tests__/UnitTestCase'
import SongListItem from './SongListItem.vue'

let row: SongRow

new class extends UnitTestCase {
  private renderComponent (song?: Song) {
    song = song ?? factory<Song>('song')

    row = {
      song,
      selected: false
    }

    return this.render(SongListItem, {
      props: {
        item: row
      }
    })
  }

  protected test () {
    it('renders', async () => {
      const song = factory<Song>('song', {
        title: 'Test Song',
        album_name: 'Test Album',
        artist_name: 'Test Artist',
        length: 1000,
        playback_state: 'Playing',
        track: 12,
        album_cover: 'https://example.com/cover.jpg',
        liked: true
      })

      const { html } = await this.renderComponent(song)
      expect(html()).toMatchSnapshot()
    })

    it('plays on double click', async () => {
      const queueMock = this.mock(queueStore, 'queueIfNotQueued')
      const playMock = this.mock(playbackService, 'play')
      this.renderComponent()

      await this.user.dblClick(screen.getByTestId('song-item'))

      expect(queueMock).toHaveBeenCalledWith(row.song)
      expect(playMock).toHaveBeenCalledWith(row.song)
    })
  }
}
