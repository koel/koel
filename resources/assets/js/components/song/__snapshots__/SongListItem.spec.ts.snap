// Vitest Snapshot v1

exports[`renders 1`] = `<div class="playing song-item" data-testid="song-item" tabindex="0"><span class="track-number"><i data-v-47e95701=""><span data-v-47e95701=""></span><span data-v-47e95701=""></span><span data-v-47e95701=""></span></i></span><span class="thumbnail"><div style="background-image: url(undefined/resources/assets/img/covers/default.svg);" class="cover" data-v-a2b2e00f=""><img alt="Test Album" src="https://example.com/cover.jpg" loading="lazy" data-v-a2b2e00f=""><a title="Pause" class="control" role="button" data-v-a2b2e00f=""><br data-testid="icon" icon="[object Object]" class="text-highlight" data-v-a2b2e00f=""></a></div></span><span class="title-artist"><span class="title text-primary">Test Song</span><span class="artist">Test Artist</span></span><span class="album">Test Album</span><span class="time">16:40</span><span class="extra"><button title="Unlike Test Song by Test Artist" type="button"><br data-testid="icon" icon="[object Object]"></button></span></div>`;
